### PR TITLE
JDK-8253733: Cleanup internal taglet API

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
This is a general cleanup of various classes related to the handling of tags and taglets in the standard doclet.

The initial motivation was to clean up static methods on `TagletWriter` that took a `TagletWriter` as a parameter. These methods are converted into instance methods, and the call sites simplified as well.

Another area of cleanup is to make clear which methods are used for inline tags and which are used for block tags.

Some dead code is removed, and the doc comments are generally cleaned up as well.  More details are given in the JBS issue.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8253733](https://bugs.openjdk.java.net/browse/JDK-8253733): Cleanup internal taglet API

### Reviewers
 * [Pavel Rappo](https://openjdk.java.net/census#prappo) (@pavelrappo - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/406/head:pull/406`
`$ git checkout pull/406`

### CODE REVIEW

**Review by pavelrappo — COMMENTED** [review]
These classes might need a copyright year update:

1. `jdk.javadoc.internal.doclets.toolkit.taglets.ValueTaglet`
2. `jdk.javadoc.internal.doclets.toolkit.taglets.SystemPropertyTaglet`
3. `jdk.javadoc.internal.doclets.toolkit.taglets.SummaryTaglet`
4. `jdk.javadoc.internal.doclets.toolkit.taglets.SeeTaglet`
5. `jdk.javadoc.internal.doclets.toolkit.taglets.ReturnTaglet`
6. `jdk.javadoc.internal.doclets.toolkit.taglets.LiteralTaglet`
7. `jdk.javadoc.internal.doclets.toolkit.taglets.IndexTaglet`
8. `jdk.javadoc.internal.doclets.toolkit.taglets.DocRootTaglet`
9. `jdk.javadoc.internal.doclets.toolkit.taglets.DeprecatedTaglet`
10. `jdk.javadoc.internal.doclets.toolkit.taglets.CodeTaglet`

It's easier to update copyright years unconditionally than to think whether or not a change amounts to a "substantial change".

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java [hunk 1]
```diff
@@ -60,186 +61,185 @@ protected TagletWriter(boolean isFirstSentence) {
...
-     * @return the output for the DocRoot inline tag.
+     * Returns the output for a {@code {@docRoot}} tag.
+     *
+     * @return the output
      */
     protected abstract Content getDocRootOutput();
 
     /**
-     * Return the deprecated tag output.
+     * Returns the output for a {@code @deprecated ...} tag.
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java:98]: In this file "ellipsis" is used for inline tags.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java [hunk 2]
```diff
@@ -60,186 +61,185 @@ protected TagletWriter(boolean isFirstSentence) {
...
+     * @param element The element that owns the doc comment
      *
-     * @param element the element to write deprecated documentation for.
-     * @return the output of the deprecated tag.
+     * @return the output
      */
     protected abstract Content deprecatedTagOutput(Element element);
 
     /**
      * Return the output for a {@code {@literal ...}} tag.
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java:107]: Add "s" to "Return" similarly to how you've done that for the rest of the methods.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java [hunk 3]
```diff
@@ -258,111 +259,119 @@ public static void genTagOutput(TagletManager tagletManager,
...
+                    tagletManager.seenTag(taglet.getName());
+                    output.add(tagletOutput);
                 }
-            }
-            if (currentOutput != null) {
-                tagletManager.seenCustomTag(taglet.getName());
-                output.add(currentOutput);
+            } catch (UnsupportedTagletOperationException e) {
+                // malformed taglet:
+                // claims to support block tags but does not provide the appropriate method.
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java:282]: This deserves a more detailed explanation: what does this claim look like? (It's not obvious from reading this excerpt.)

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java [hunk 4]
```diff
@@ -258,111 +259,119 @@ public static void genTagOutput(TagletManager tagletManager,
...
+     * @return the generated content
      */
-    public abstract Content commentTagsToOutput(DocTree holderTag,
-        Element holder, List<? extends DocTree> tags, boolean isFirstSentence);
+    public abstract Content commentTagsToOutput(Element element, DocTree holder,
+                                                List<? extends DocTree> trees, boolean isFirstSentence);
 
     /**
-     * @return an instance of the configuration used for this doclet.
+     * Returns  an instance of the configuration used for this doclet.
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java:372]: Remove extra whitespace in "Returns  an".

**Review by pavelrappo — CHANGES_REQUESTED** [review]

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java (hunk 1 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java:98]: OK.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java (hunk 2 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java:107]: Good catch

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java (hunk 3 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java:282]: Fixed

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java (hunk 4 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java:372]: 👍

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SimpleTaglet.java [hunk 5]
```diff
@@ -178,17 +178,17 @@ public void inherit(DocFinder.Input input, DocFinder.Output output) {
     }
 
     @Override
-    public Content getTagletOutput(Element element, DocTree tag, TagletWriter writer) {
-        return header == null || tag == null ? null : writer.simpleTagOutput(element, tag, header);
+    public Content getInlineTagOutput(Element element, DocTree tag, TagletWriter writer) {
+        return null;
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SimpleTaglet.java:182]: A better implementation would be remove this method, and rely on the underlying default impl to throw an exception; I'll change code to do that.  The reason is that `SimpleTag` is only used for block tags, and this method is is (now) clearly for inline tags.

This sort of confusion in the original code was significant contributing factor in replacing the one name for two overloaded methods with two separate names for the different usages.

**Comment by jonathan-gibbons** [comment]
@pavelrappo , thanks for the list of files that need copyright updates; I've not yet figured out how to update the scripts that I used to use to update files, and I used to do it after review and before push, to minimize noise in the review. Old habits die hard.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java [hunk 6]
```diff
@@ -279,7 +279,8 @@ public Content getBlockTagOutput(TagletManager tagletManager,
                 }
             } catch (UnsupportedTagletOperationException e) {
                 // malformed taglet:
-                // claims to support block tags but does not provide the appropriate method.
+                // claims to support block tags (see Taglet.isBlockTag) but does not provide the
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java:282]: This version does a better job of explaining your assumptions. Thanks.

(You don't need to address the below comment.)

I think this method could benefit from an assertion or an unconditional check that would guard against calling this method with a taglet list containing a non-block taglet. That is, a taglet `t` such that `t.isBlockTag() == false`.

This is a bit different from `TagletWriter#getInlineTagOutput` where the check is built into the method; `TagletWriter#getBlockTagOutput` relies on a caller to pass the correct list.

**Review by pavelrappo — APPROVED** [review]

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java (hunk 6 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java:282]: Interesting question, worth considering ...

This is an internal API, so is only called from internal sites, with one of two lists supplied by the TagletManager: standard block tags, or serial form tags (the latter is why the list comes in as an argument and is not provided internally.)

But given the list comes in as an argument, that suggests a reasonable update is to verify that all items in the argument list are block tags, and to throw ILA if any are not.

**Review by pavelrappo — APPROVED** [review]
Perfect, thanks!

**Pull request closed without merging**